File: src/lib/email/providers/smtp.ts
import { isTauri } from "@/lib/runtime";
import type {
  EmailProvider,
  OutgoingEmail,
  SendResult,
  SmtpConfig,
} from "@/lib/email/types";

/**
 * SMTP provider. The actual network send happens in the Tauri Rust backend
 * (using the `lettre` crate) via the `smtp_test` and `smtp_send` commands —
 * browsers cannot open raw SMTP sockets. When running outside Tauri (the dev
 * server), sending is unavailable and returns a clear, safe error rather than
 * pretending to send.
 */
async function invokeTauri<T>(cmd: string, args: Record<string, unknown>): Promise<T> {
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<T>(cmd, args);
}

const NOT_AVAILABLE: SendResult = {
  ok: false,
  error:
    "Sending requires the desktop app. SMTP is not available in the browser dev preview.",
};

export const smtpProvider: EmailProvider = {
  id: "smtp",
  label: "SMTP",

  async testConnection(config: SmtpConfig, password: string): Promise<SendResult> {
    if (!isTauri()) return NOT_AVAILABLE;
    try {
      await invokeTauri<void>("smtp_test", { config: toRustConfig(config, password) });
      return { ok: true };
    } catch (err) {
      return { ok: false, error: String(err) };
    }
  },

  async sendOne(
    config: SmtpConfig,
    password: string,
    email: OutgoingEmail,
  ): Promise<SendResult> {
    if (!isTauri()) return NOT_AVAILABLE;
    try {
      await invokeTauri<void>("smtp_send", {
        config: toRustConfig(config, password),
        email,
      });
      return { ok: true };
    } catch (err) {
      return { ok: false, error: String(err) };
    }
  },
};

/** Shape the Rust command expects (snake_case-friendly, includes password). */
function toRustConfig(config: SmtpConfig, password: string) {
  return {
    host: config.host,
    port: config.port,
    username: config.username,
    password,
    senderEmail: config.senderEmail,
    senderName: config.senderName,
    encryption: config.encryption,
  };
}

const PROVIDERS: Record<string, EmailProvider> = {
  [smtpProvider.id]: smtpProvider,
};

export function getEmailProvider(id: string): EmailProvider {
  const p = PROVIDERS[id];
  if (!p) throw new Error(`Unknown email provider: ${id}`);
  return p;
}

File: src-tauri/src/smtp.rs
//! SMTP test sending via the `lettre` crate.
//!
//! Browsers cannot open raw SMTP sockets, so the actual send happens here in
//! the Rust backend. Two commands are exposed: `smtp_test` (verify the
//! connection/credentials) and `smtp_send` (send a single message). Bulk
//! sending is intentionally NOT implemented in the MVP.

use lettre::message::{header::ContentType, Mailbox, MultiPart, SinglePart};
use lettre::transport::smtp::authentication::Credentials;
use lettre::transport::smtp::client::{Tls, TlsParameters};
use lettre::{AsyncSmtpTransport, AsyncTransport, Message, Tokio1Executor};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SmtpConfig {
    pub host: String,
    pub port: u16,
    pub username: String,
    pub password: String,
    pub sender_email: String,
    pub sender_name: String,
    /// "ssl_tls" | "starttls" | "none"
    pub encryption: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutgoingEmail {
    pub to: String,
    pub from_name: String,
    pub from_email: String,
    pub subject: String,
    pub text: String,
    pub html: String,
}

fn build_transport(config: &SmtpConfig) -> Result<AsyncSmtpTransport<Tokio1Executor>, String> {
    let creds = Credentials::new(config.username.clone(), config.password.clone());

    let builder = match config.encryption.as_str() {
        "ssl_tls" => {
            let tls = TlsParameters::new(config.host.clone())
                .map_err(|e| format!("TLS setup failed: {e}"))?;
            AsyncSmtpTransport::<Tokio1Executor>::builder_dangerous(&config.host)
                .tls(Tls::Wrapper(tls))
        }
        "starttls" => {
            let tls = TlsParameters::new(config.host.clone())
                .map_err(|e| format!("TLS setup failed: {e}"))?;
            AsyncSmtpTransport::<Tokio1Executor>::builder_dangerous(&config.host)
                .tls(Tls::Required(tls))
        }
        _ => AsyncSmtpTransport::<Tokio1Executor>::builder_dangerous(&config.host),
    };

    Ok(builder.port(config.port).credentials(creds).build())
}

#[tauri::command]
pub async fn smtp_test(config: SmtpConfig) -> Result<(), String> {
    let transport = build_transport(&config)?;
    match transport.test_connection().await {
        Ok(true) => Ok(()),
        Ok(false) => Err("Server did not accept the connection.".into()),
        Err(e) => Err(format!("Connection failed: {e}")),
    }
}

#[tauri::command]
pub async fn smtp_send(config: SmtpConfig, email: OutgoingEmail) -> Result<(), String> {
    let from: Mailbox = format!("{} <{}>", email.from_name, email.from_email)
        .parse()
        .map_err(|_| "Invalid sender address".to_string())?;
    let to: Mailbox = email
        .to
        .parse()
        .map_err(|_| "Invalid recipient address".to_string())?;

    let message = Message::builder()
        .from(from)
        .to(to)
        .subject(email.subject)
        .multipart(
            MultiPart::alternative()
                .singlepart(
                    SinglePart::builder()
                        .header(ContentType::TEXT_PLAIN)
                        .body(email.text),
                )
                .singlepart(
                    SinglePart::builder()
                        .header(ContentType::TEXT_HTML)
                        .body(email.html),
                ),
        )
        .map_err(|e| format!("Failed to build message: {e}"))?;

    let transport = build_transport(&config)?;
    transport
        .send(message)
        .await
        .map_err(|e| format!("Send failed: {e}"))?;
    Ok(())
}
